src/System.Threading.Channels/src/System/Threading/Channels/AsyncOperation.cs

```diff
@@ -0,0 +1,370 @@
+        protected static readonly Action<object> s_completedSentinel = new Action<object>(s => Debug.Fail($"{nameof(ResettableValueTaskSource)}.{nameof(s_completedSentinel)} invoked."));
+
+        /// <summary>Throws an exception indicating that the operation's result was accessed before the operation completed.</summary>
+        protected static void ThrowIncompleteOperationException() =>
+            throw new InvalidOperationException(SR.InvalidOperation_IncompleteAsyncOperation);
+
+        /// <summary>Throws an exception indicating that multiple continuations can't be set for the same operation.</summary>
+        protected static void ThrowMultipleContinuations() =>
+            throw new InvalidOperationException(SR.InvalidOperation_MultipleContinuations);
+
+        /// <summary>Throws an exception indicating that the operation was used after it was supposed to be used.</summary>
+        protected static void ThrowIncorrectCurrentIdException() =>
+            throw new InvalidOperationException(SR.InvalidOperation_IncorrectToken);
+
+        /// <summary>Describes states the operation can be in.</summary>
+        public enum States
+        {
+            /// <summary>The operation has been assigned an owner.  No one else can use it.</summary>
+            Owned = 0,
+            /// <summary>Completion has been reserved.  Only the reserver is allowed to complete it.</summary>
+            CompletionReserved = 1,
+            /// <summary>The operation has completed and has had its result or error stored.</summary>
+            CompletionSet = 2,
+            /// <summary>The operation's result/error has been retrieved.  It's available for reuse.</summary>
+            Released = 3
+        }
+    }
+
+    internal abstract class ResettableValueTaskSource<T> : ResettableValueTaskSource, IValueTaskSource, IValueTaskSource<T>
+    {
+        private volatile int _state = (int)States.Owned;
+        private T _result;
+        private ExceptionDispatchInfo _error;
+        private Action<object> _continuation;
+        private object _continuationState;
+        private object _schedulingContext;
+        private ExecutionContext _executionContext;
+        private short _currentId;
+
+        public ValueTask CreateNonGenericValueTask() => new ValueTask(this, _currentId);
+        public ValueTask<T> CreateGenericValueTask() => new ValueTask<T>(this, _currentId);
+
+        public bool RunContinutationsAsynchronously { get; protected set; }
+        public ValueTaskSourceStatus GetStatus(short token)
+        {
+            if (_currentId != token)
+            {
+                throw new InvalidOperationException();
+            }
+
+            switch ((States)_state)
+            {
+                case States.Owned:
+                case States.CompletionReserved:
+                    return ValueTaskSourceStatus.Pending;
+
+                case States.CompletionSet:
+                case States.Released:
```

Comment:
Why isn't Released an error?  It seems like there's nothing good that can come from accessing the object in this state, right?

Reply:
I can probably make it one now.  Previously when we just had IsCompleted, I had to roll them together based on an implementation detail elsewhere, but I should be able to address this now.  I will if I can.